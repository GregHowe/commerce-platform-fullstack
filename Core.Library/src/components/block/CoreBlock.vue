<template>
	<component
		:is="blockTag"
		:id="blockId"
		:target="linkTarget"
		:class="blockClasses"
		:href="settings.linkInternal ? '' : linkUrl"
		:to="settings.linkInternal ? '' : linkUrl"
	>
		<component
			:is="blockComponent"
			:settings="settings"
			:nested-level="nestedLevel"
			:index="index"
			:is-editable="isEditable"
			:data-site="dataSite"
			@select="select"
			@insert="insert"
		/>
		<slot name="controls" />
		<component
			:is="'style'"
			v-if="blockStyle"
		>
			{{ blockStyle }}
		</component>
	</component>
</template>

<script>
import { sitePath } from "@libraryHelpers/dataComponents.js";

export default {
	name: "CoreBlock",
	props: {
		index: {
			type: Number,
			default: 0,
		},
		isEditable: {
			type: Boolean,
			default: false,
		},
		nestedLevel: {
			type: Number,
			default: 0,
		},
		settings: {
			type: Object,
			required: true,
		},
	},
	computed: {
		dataSite() {
			// this is the site object used in both the generator and the builder
			return sitePath(this.$store);
		},

		blockTag() {
			if (this.settings?.linkUrl) {
				if (this.settings?.linkInternal) {
					return "NuxtLink";
				}
				return "a";
			}
			return this.settings?.linkUrl ? "a" : "div";
		},
		linkUrl() {
			return this.settings?.linkUrl;
		},
		linkTarget() {
			return this.settings.linkInternal ? "_self" : "_blank";
		},
		blockComponent() {
			switch (this.settings.type) {
				case "button":
					return "CoreBlockButton";
				case "breadcrumbs":
					return "CoreBlockBreadcrumbs";
				case "contact-card":
					return "CoreBlockContactCard";
				case "contact-hero":
					return "CoreBlockContactHero";
				case "columns":
					return "CoreBlockColumns";
				case "container":
				case "section":
				case "accordion":
				case "toggle":
				case "tabs":
					return "CoreBlockContainer";
				case "content-card":
					return "CoreBlockContentCard";
				case "divider":
					return "CoreBlockDivider";
				case "file":
					return "CoreBlockFile";
				case "grid":
					return "CoreBlockGrid";
				case "hero":
					return "CoreBlockHero";
				case "html": 
					return "CoreBlockHtml";
				case "icon":
					return "CoreBlockIcon";
				case "image":
					return "CoreBlockImage";
				case "media":
					return "CoreBlockMedia";
				case "rows":
					return "CoreBlockRows";
				case "social-links":
					return "CoreBlockSocialLinks";
				case "text":
					return "CoreBlockText";
				case "video":
					return "CoreBlockVideo";
				case "pencil-banner":
					return "CoreBlockPencilBanner";
				case "rich-text":
					return "CoreBlockRichText";
				case "form":
					return "CoreBlockForm";
				case "calendly":
					return "CoreBlockCalendly";
				case "map":
					return "CoreBlockMap";
			}
			return "CoreBlockUnknown";
		},
		blockId() {
			return this.settings.id;
		},
		blockStyle() {
			const styleSetting = this.settings.style || "";
			return [styleSetting]
				.join("\n")
				.replaceAll("#id", `#${this.blockId}`);
		},
		blockType() {
			return this.settings.type;
		},
		blockVariants() {
			const variants = [];
			if (this.settings.variants) {
				for (const [variantKey, variantValue] of Object.entries(
					this.settings.variants
				)) {
					// go through every variant saved that is not of the "_custom" key
					if (variantKey !== "_custom") {
						// variants defined without options are just toggles
						// and if they are true, then apply the variant key
						if (variantValue === true) {
							variants.push(variantKey);
						}
						// variants defined with specific string value options
						// should apply the selected value
						else {
							variants.push(variantValue);
						}
					}
				}
			}
			if (!variants.length) {
				variants.push("base");
			}
			return variants;
		},
		blockClasses() {
			const classList = ["block", `block_${this.blockType}`];
			this.blockVariants.forEach((variant) => {
				if (variant) {
					classList.push(`block_${this.blockType}_${variant}`);
				}
			});
			// anything added to the _custom key is a string w/ space delimited custom classes
			if (this.settings.variants?._custom) {
				classList.push(this.settings.variants._custom.split(" "));
			}
			return classList;
		},
	},
	methods: {
		insert(payload) {
			// relay it up to the forge control block
			this.$emit("insert", payload);
		},
		select(payload) {
			// relay it up to the forge control block
			this.$emit("select", payload);
		},
	},
};
</script>

<style lang="scss">
// styles universal to blocks (will refactor into CoreLibrary)

.block {
	min-height: 1rem;
	min-width: 1rem;
	z-index: 101;
	.block__body {
		z-index: 201;
	}
}

.block {
	position: relative;
	width: 100%;
	height: 100%;
	.block {
		background: transparent;
	}
	.block__body {
		position: relative;
		margin: 0 auto;
		padding: var(--_core__container_padding-top)
			var(--_core__container_padding-right)
			var(--_core__container_padding-bottom)
			var(--_core__container_padding-left);
		background: transparent;
		color: inherit;
		font-family: inherit;
		text-align: var(--_core__body_text-align);
		justify-content: var(--_core__body_justify-content);
		max-width: 100%;
		.block__container {
			max-width: var(--_core__container_max-width);
			width: 100%;
			margin-right: auto;
			margin-left: auto;
		}
	}

	--_core__list_margin-left: auto;
	--_core__list_margin-right: auto;
	&[class*="_align-left"] {
		.block__body {
			--_core__body_text-align: left;
			--_core__body_justify-content: flex-start;
			--_core__list_margin-left: 0;
			--_core__list_margin-right: auto;
		}
	}
	&[class*="_image-align-left"] {
		.block__body {
			.block__section_content {
				display: flex;
				justify-content: flex-start;
				img {
					margin: 0;
				}
			}
		}
	}
	&[class*="_image-align-center"] {
		.block__body {
			.block__section_content {
				display: flex;
				justify-content: center;
				img {
					margin: 0 auto;
				}
			}
		}
	}
	&[class*="_image-align-right"] {
		.block__body {
			.block__section_content {
				width: 100%;
				display: flex;
				justify-content: flex-end;
				img {
					margin: 0;
				}
			}
		}
	}

	&[class*="_align-center"] {
		.block__body {
			--_core__body_text-align: center;
			--_core__body_justify-content: center;
			--_core__list_margin-left: auto;
			--_core__list_margin-right: auto;
		}
	}

	&[class*="_align-right"] {
		.block__body {
			--_core__body_text-align: right;
			--_core__body_justify-content: flex-end;
			--_core__list_margin-left: auto;
			--_core__list_margin-right: 0;
		}
	}

	/*
	each block resets the default padding within
	otherwise nested blocks would inherit the padding of their parents
	*/
	--_core__container_padding-top: 0;
	--_core__container_padding-right: 0;
	--_core__container_padding-bottom: 0;
	--_core__container_padding-left: 0;

	/*

	padding related classes that can be applied to any block
	---
	every block has the variants 'padding-top', 'padding-bottom', and 'padding-horizontal'
	which will apply classes ending with things like 'pt-small' (padding-top small)
	so changing the value of the theme's variables inside each component
	to match whatever variant was selected will adjust the spacing on the fly

	*/
	&[class*="_pt-none"] {
		--_core__container_padding-top: 0;
	}
	&[class*="_pt-small"] {
		--_core__container_padding-top: 1rem;
	}
	&[class*="_pt-medium"] {
		--_core__container_padding-top: 2rem;
	}
	&[class*="_pt-large"] {
		--_core__container_padding-top: 3rem;
	}
	&[class*="_pt-huge"] {
		--_core__container_padding-top: 5rem;
	}
	&[class*="_pb-none"] {
		--_core__container_padding-bottom: 0;
	}
	&[class*="_pb-small"] {
		--_core__container_padding-bottom: 1rem;
	}
	&[class*="_pb-medium"] {
		--_core__container_padding-bottom: 2rem;
	}
	&[class*="_pb-large"] {
		--_core__container_padding-bottom: 3rem;
	}
	&[class*="_pb-huge"] {
		--_core__container_padding-bottom: 5rem;
	}
	&[class*="_px-none"] {
		--_core__container_padding-right: 0;
		--_core__container_padding-left: 0;
	}
	&[class*="_px-small"] {
		--_core__container_padding-right: 1rem;
		--_core__container_padding-left: 1rem;
	}
	&[class*="_px-medium"] {
		--_core__container_padding-right: 2rem;
		--_core__container_padding-left: 2rem;
	}
	&[class*="_px-large"] {
		--_core__container_padding-right: 3rem;
		--_core__container_padding-left: 3rem;
	}
	&[class*="_px-huge"] {
		--_core__container_padding-right: 5rem;
		--_core__container_padding-left: 5rem;
	}
	&[class*="_p-none"] {
		--_core__container_padding: 0;
	}
	&[class*="_p-small"] {
		--_core__container_padding: 1rem;
	}
	&[class*="_p-medium"] {
		--_core__container_padding: 2rem;
	}
	&[class*="_p-large"] {
		--_core__container_padding: 3rem;
	}
	&[class*="_p-huge"] {
		--_core__container_padding: 5rem;
	}
}
</style>
