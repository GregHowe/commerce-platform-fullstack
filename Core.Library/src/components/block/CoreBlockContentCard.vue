<template>
	<div class="block__body block__content-card">
		<div class="container">
			<CoreBlock
				v-if="mediaSrc"
				:settings="{
					type: 'media',
					src: mediaSrc,
					height: 'auto',
					width: 'auto',
				}"
			/>

			<div class="block__section_text">
				<CoreBlock
					v-if="overlineText"
					:settings="{
						type: 'text',
						text: overlineText,
						tag: 'h4',
						variants: {
							size: 'overline',
						},
					}"
				/>
				<CoreBlock
					v-if="headlineText"
					:settings="{
						type: 'text',
						text: headlineText,
						tag: 'h5',
						variants: {
							size: 'headline',
						},
					}"
				/>
				<CoreBlock
					v-if="richText"
					:settings="{
						type: 'rich-text',
						richText,
					}"
				/>
				<div
					v-if="hasButtons"
					class="block__section_btns"
				>
					<CoreBlock
						v-if="hasPrimaryButton"
						:settings="{
							type: 'button',
							url: settings.primaryUrl,
							label: settings.primaryLabel,
							newTab: settings.primaryBtnNewTab,
							variants: { style: 'primary' },
						}"
					/>
					<CoreBlock
						v-if="hasSecondaryButton"
						:settings="{
							type: 'button',
							url: settings.secondaryUrl,
							label: settings.secondaryLabel,
							newTab: settings.secondaryBtnNewTab,
							icon: 'right-arrow',
							variants: {
								style: 'text',
								outline: true,
								'icon-placement': 'icon-right',
							},
						}"
					/>
				</div>

				<CoreBlock
					v-if="disclosureText"
					:settings="{
						type: 'rich-text',
						richText: disclosureText,
						variants: {
							size: 'disclosure',
						},
					}"
				/>
			</div>
		</div>
	</div>
</template>

<script>
import { renderData } from "@libraryHelpers/dataComponents.js";

export default {
	name: "CoreBlockContentCard",
	props: {
		settings: {
			type: Object,
			required: true,
		},
		dataSite: {
			type: Object,
			required: true,
		},
	},
	computed: {
		overlineText() {
			return this.settings?.overline
				? renderData(this.settings?.overline, this.dataSite)
				: null;
		},
		headlineText() {
			return this.settings?.headline
				? renderData(this.settings?.headline, this.dataSite)
				: null;
		},
		richText() {
			return this.settings?.richText
				? renderData(this.settings?.richText, this.dataSite)
				: null;
		},
		disclosureText() {
			return this.settings?.disclosure
				? renderData(this.settings.disclosure, this.dataSite)
				: null;
		},
		mediaSrc() {
			return this.settings?.media || null;
		},
		hasButtons() {
			return this.hasPrimaryButton || this.hasSecondaryButton;
		},
		hasPrimaryButton() {
			return this.settings?.primaryButton || false;
		},
		hasSecondaryButton() {
			return this.settings?.secondaryButton || false;
		},
	},
};
</script>

<style lang="scss">
// .block.block--selected-deeply .block.block--selected {
//   outline: 0 solid transparent !important;
// }
.block_container_slim {
	.block_content-card {
		.container {
			border: 1px solid var(--_core__section_border-color);
		}
		.block_text_headline h5 {
			min-height: 100px;
		}
	}
}
.block_content-card {
	padding: 0 !important;
	align-self: flex-start;

	.block__body.block__content-card {
		padding-left: 0;
	}
	.container {
		padding: var(--_core__content_card-padding);
		// padding: 20px 20px;
		word-wrap: break-word;
	}
	.block_image .block__body .block__section_content img {
		object-fit: cover;
	}

	.block_text_title,
	.block_text_headline {
		// margin-top: 10px;
	}

	.block__body {
		display: flex;
		flex-direction: column;
		text-align: var(--_core__body_text-align);
		justify-content: var(--_core__body_justify-content);

		.block_text .block__body {
			color: var(--_core__content_card-font-color);
		}
		.block__section_text {
			display: flex;
			flex-direction: column;
			padding: var(--_core__content_card-text-padding);
			background-color: var(--_core__content_card-background-color);
			color: var(--_core__content_card-font-color);
			.block__body {
				color: var(--_core__content_card-font-color);
			}
			.block__section_btns {
				padding-bottom: 0;
			}
		}
		.block__section_btns {
			display: flex;
			flex-direction: row;
			align-items: center;
			justify-content: var(--_core__body_justify-content);
			padding-bottom: 1rem;
			width: 100%;
			& > .block {
				width: auto;
				margin: 0 0.2rem;
				&:first-child {
					margin-left: 0;
				}
				&:last-child {
					margin-right: 0;
				}
			}
		}
		.primary-button {
			a {
				border-style: solid;
				border-radius: 0.3rem;
				border-color: var(--_core__button_primary-border);
			}
		}
		.secondary-btn {
			a {
				background-color: var(--_core__button_secondary-border);
				border-style: solid;
				border-color: var(--_core__button_primary-border);
				color: var(--_core__button_secondary-color);
				border-radius: 0.3rem;
				inline-size: max-content;
			}
		}
		.block_button.block_button_text .button {
			color: var(--_core__content_card-link-color);
		}
		.block__section_overline {
			color: var(--_core__overline_color);
			font-size: var(--_core__overline_font-size);
			text-transform: var(--_core__overline_text-transform);
			font-family: var(--_core__overline_font-family);
			font-weight: 400;
			margin: 0.4rem 0;
		}
		.block__section_headline {
			font-size: var(--_core__header_font-size);
			line-height: var(--_core__header_line-height);
			color: var(--_core__header_color);
			text-transform: var(--_core__header_text-transform);
			font-family: var(--_core__header_font-family);
			padding-bottom: 1rem;
		}
		.block__section_disclosure {
			font-size: var(--_core__disclosure_font-size);
			color: var(--_core__disclosure_color);
			font-style: var(--_core__disclosure_font-style);
			line-height: var(--_core__disclosure_line-height);
		}
	}

	/* move these up to be variants available to every type of block */
	&.block_content-card_bg-color {
		--_core__container_background: var(--core__color_primary);
		--_core__container_color: var(--core__color_light-alt);
	}

	&.block_content-card_bg-dark {
		--_core__container_background: var(--core__color_dark);
		--_core__container_color: var(--core__color_light-alt);
	}

	&.block_content-card_bg-light {
		--_core__container_background: var(--core__color_light);
		--_core__container_color: var(--core__color_dark-alt);
	}

	&.block_content-card_image-right {
		@include breakpoint_up(lg) {
			.block__body {
				display: flex;
				flex-direction: row-reverse;
				align-items: center;
			}
		}
	}
	&.block_content-card_image-left {
		@include breakpoint_up(lg) {
			.block__body {
				display: flex;
				flex-direction: row;
				align-items: center;
			}
		}
	}
}
</style>
