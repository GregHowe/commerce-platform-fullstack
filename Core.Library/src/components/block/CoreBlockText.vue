<template>
	<component
		:is="blockTag"
		v-if="blockContent"
		class="block__body"
	>
		{{ blockContent }}
	</component>
</template>

<script>
export default {
	name: "CoreBlockText",
	props: {
		settings: {
			type: Object,
			required: true,
		},
	},
	computed: {
		blockContent() {
			return this.settings?.text || null;
		},
		blockTag() {
			return this.settings?.tag || "div";
		},
	},
};
</script>

<style lang="scss">
.block_text {
	.block__body {
		font-size: var(--_core__container_font-size);
		text-transform: none;
		font-weight: 400;
		font-style: normal;
		color: var(--_core__container_color);
		font-family: var(--_core__container_font-family);
		text-align: var(_core__container_text-align);
		margin: 0.4rem 0;
	}

	&.block_text_overline {
		.block__body {
			color: var(--_core__overline_color);
			font-size: var(--_core__overline_font-size);
			text-transform: var(--_core__overline_text-transform);
			font-family: var(--_core__overline_font-family);
		}
	}

	&.block_text_title,
	&.block_text_headline {
		.block__body {
			line-height: var(--_core__heading_line-height);
			// color: var(--_core__heading_color);
			text-transform: var(--_core__heading_text-transform);
			font-family: var(--_core__heading_font-family);
			font-weight: 500;			
		}
		h1.block__body {
			font-size: var(--_core__heading_font-size--h1);
			font-family: 	var(--_core__heading_font-family--h1);
		}
		h2.block__body {
			font-size: var(--_core__heading_font-size--h2);
			font-family: 	var(--_core__heading_font-family--h2);
			color: var(	--_core__heading_color--h2);
		}
		h3.block__body {
			font-size: var(--_core__heading_font-size--h3);
			font-family: 	var(--_core__heading_font-family--h3);
		}
		h4.block__body {
			font-size: var(--_core__heading_font-size--h4);
			font-family: 	var(--_core__heading_font-family--h4);
		}
		h5.block__body {
			font-size: 22px;
			font-family: 	var(--_core__heading_font-family--h5);
			text-transform: var(--_core__heading_text-transform--h5);
			font-weight: 600;
		}
	}

	&.block_text_subtitle,
	&.block_text_subline {
		.block__body {
			font-size: var(--_core__subline_font-size);
			line-height: var(--_core__subline_line-height);
			color: var(--_core__subline_color);
			text-transform: var(--_core__subline_text-transform);
			font-family: var(--_core__subline_font-family);
			margin: 0.4rem 0 0.8rem;
		}
	}

	&.block_text_footnote,
	&.block_text_disclosure {
		.block__body {
			font-size: var(--_core__disclosure_font-size);
			color: var(--_core__disclosure_color);
			font-style: var(--_core__disclosure_font-style);
			line-height: var(--_core__disclosure_line-height);
		}
	}

	&.block_text_bold {
		.block__body {
			font-weight: 600;
		}
	}

	&.block_text_italic {
		.block__body {
			font-style: italic;
		}
	}

	&.block_text_headline + .block_rich-text,
	&.block_text_headline + .block_text {
		// margin-top: 0.6em;
		margin-bottom: 0.6em;
	}
}
</style>
