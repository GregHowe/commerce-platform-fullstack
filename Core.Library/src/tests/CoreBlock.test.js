import CoreBlock from "../components/block/CoreBlock";
import "@testing-library/jest-dom";
import { render } from "@testing-library/vue";
import CoreBlockText from "../components/block/CoreBlockText";
import CoreBlockContentCard from "../components/block/CoreBlockContentCard";

const text = "hahaha hahahaha HAHAHAHA";

const renderComponent = async (props = {}) => {
	return render(CoreBlock, {
		props,
		components: { CoreBlockText, CoreBlockContentCard },
	});
};
describe("CoreBlock text component", () => {
	test("text component is shown if settings type is text", async () => {
		const { container } = await renderComponent({
			settings: { type: "text", text },
		});
		expect(container).toMatchSnapshot();
	});

	test("text component size variants adjust classes", async () => {
		const { container } = await renderComponent({
			settings: {
				type: "text",
				text,
				variants: {
					align: "center",
					size: "title",
					style: "italic",
					weight: "bold",
				},
			},
		});

		expect(container).toMatchSnapshot();
	});
});

describe("CoreBlock rich-text component", () => {
	test("rich text component is shown if settings type is rich-text", async () => {
		const { container } = await renderComponent({
			settings: { type: "rich-text", richText: text },
		});
		expect(container).toMatchSnapshot();
	});
});
