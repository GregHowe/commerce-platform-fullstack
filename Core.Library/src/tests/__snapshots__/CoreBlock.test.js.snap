// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreBlock rich-text component rich text component is shown if settings type is rich-text 1`] = `
<div>
  <div
    class="block block_rich-text block_rich-text_base"
    target="_blank"
  >
    <div
      class="block__body"
      index="0"
      nested-level="0"
    >
      hahaha hahahaha HAHAHAHA
    </div>
      
    <!---->
  </div>
</div>
`;

exports[`CoreBlock text component text component is shown if settings type is text 1`] = `
<div>
  <div
    class="block block_text block_text_base"
    target="_blank"
  >
    <div
      class="block__body"
      index="0"
      nested-level="0"
    >
      
	hahaha hahahaha HAHAHAHA

    </div>
      
    <!---->
  </div>
</div>
`;

exports[`CoreBlock text component text component size variants adjust classes 1`] = `
<div>
  <div
    class="block block_text block_text_center block_text_title block_text_italic block_text_bold"
    target="_blank"
  >
    <div
      class="block__body"
      index="0"
      nested-level="0"
    >
      
	hahaha hahahaha HAHAHAHA

    </div>
      
    <!---->
  </div>
</div>
`;
